ext/TensorKitChainRulesCoreExt/utility.jl

```diff
@@ -29,3 +29,15 @@ function (::ProjectTo{T1})(x::T2) where {S,N1,N2,T1<:AbstractTensorMap{<:Any,S,N
     end
     return y
 end
+
+function (::ProjectTo{T1})(x::T2) where {S,NumType,StorType,
+                                         T1<:DiagonalTensorMap{NumType,S,StorType},
+                                         T2<:AbstractTensorMap{<:Any,S,1,1}}
+    T1 === T2 && return x
+    y = DiagonalTensorMap{NumType,S,StorType}(undef, space(x, 1))
```

Comment:
```suggestion
function (::ProjectTo{T1})(x::T2) where {T1<:DiagonalTensorMap,T2<:AbstractTensorMap}
    T1 === T2 && return x
    y = DiagonalTensorMap{scalartype(T1),spacetype(T1),storagetype(T1)}(undef, space(x, 1))
```
In general, I prefer to not use too many parametric types in the signature, and rather use the functions to obtain them. I feel like that's a bit more flexible, and also somewhat easier to read.